packages/solo/src/chain-cosmos-sdk.js

```diff
@@ -381,137 +330,210 @@ ${chainID} chain does not yet know of address ${myAddr}${adviseEgress(myAddr)}
       }
 
-      const args = [
-        'tx',
-        'swingset',
-        'deliver',
+      args.push(
         '--keyring-backend=test',
-        `@${tmpInfo.path}`, // Deliver message over file, as it could be big.
         '--gas=auto',
-        '--gas-adjustment=1.2',
+        '--gas-adjustment=1.3',
         '--from=ag-solo',
-        '--broadcast-mode=block', // Don't return until committed.
         '--yes',
-      ];
-
-      // If we send two messages back-to-back too quickly, the second one
-      // may use the same seqnum as the first, so it will be rejected by the
-      // signature-checking auth/ante handler on the chain. Therefore, we need
-      // to queue our deliver() calls and not allow one to proceed until the
-      // others have finished.
-      const qret = await queuedHelper(
-        'deliver',
-        undefined, // allow the queue to grow unbounded, and never cancel deliveries
-        args,
-        ret => {
-          const { stderr, stdout } = ret;
-          const errMsg = stderr.trimRight();
-          if (errMsg) {
-            log.error(errMsg);
+      );
+
+      // Use the feeAccount for any fees.
+      if (feeAccountAddr) {
+        args.push(`--fee-account=${feeAccountAddr}`);
+      }
+
+      // We just try a single delivery per block.
+      let retry = true;
+      while (retry) {
+        retry = false;
+        // eslint-disable-next-line no-await-in-loop
+        const { stderr, stdout } = await runHelper([
+          ...args,
+          `--sequence=${sequenceNumber}`,
+        ]);
+
+        const errMsg = stderr.trimRight();
+        if (errMsg) {
+          const seqMatch = errMsg.match(
+            /account sequence mismatch, expected (\d+),/,
+          );
+          if (seqMatch) {
+            // Try to resync our sequence number before sending more.
+            console.info(
+              `Resynchronizing ${GCI} sequence number from ${sequenceNumber} to ${seqMatch[1]}`,
+            );
+            sequenceNumber = BigInt(seqMatch[1]);
+            retry = true;
+          } else {
+            console.error(errMsg);
           }
-          log.debug(`helper said: ${stdout}`);
+        }
+
+        if (!retry) {
+          console.debug(`helper said: ${stdout}`);
           const out = JSON.parse(stdout);
-          if (Number(out.height) > 0) {
-            // We submitted the transaction successfully.
-            return {};
-          }
-          assert.fail(X`Unexpected output: ${stdout.trimRight()}`);
-        },
-        undefined,
-        {}, // defaultIfCancelled
-      );
-      return qret;
+
+          assert.equal(
+            parseInt(out.code, 10),
+            0,
+            X`Unexpected output: ${stdout.trimRight()}`,
+          );
+
+          // We submitted the transaction successfully.
+          sequenceNumber += 1n;
+        }
+      }
+    } catch (e) {
+      // Put back the deliveries we tried to make.
+      messagePool = messagePool.concat(messages);
```

Comment:
Would this cause messages to be delivered out-of-order sometimes?

I'm thinking message 1+2 get added to the pool, then the delivery is attempted, and while the attempt is running someone adds messages 3+4 to the pool. Then the delivery finishes failing, so 1+2 are `concat`'ed back in, and now the pool looks like 3+4+1+2, and the next time we deliver things, they get sent out of order.

What swingset (in particular the mailbox device) really wants is `{ messages: [[1,m1],[2,m2],[3,m3]], ackNum: N }`. I *think* the mailbox device can handle them appearing out of order, but it's probably better to meet its expectations rather than rely upon that.

Certainly if we introduce size limits on the messages, we'll need to ensure that we send m1+m2 before m3+m4, so keeping the message pool sorted will probably avoid problems at that points.